SteamKit2/SteamKit2/Steam/Handlers/SteamMatchmaking/Lobby.cs

```diff
@@ -0,0 +1,402 @@
+            /// </summary>
+            public SteamID OwnerSteamID { get; }
+
+            /// <summary>
+            /// The metadata of the lobby; string key-value pairs.
+            /// </summary>
+            public IReadOnlyDictionary<string, string> Metadata { get; }
+
+            /// <summary>
+            /// The maximum number of members that can occupy the lobby.
+            /// </summary>
+            public int MaxMembers { get; }
+
+            /// <summary>
+            /// The number of members that are currently occupying the lobby.
+            /// </summary>
+            public int NumMembers { get; }
+
+            /// <summary>
+            /// A list of lobby members. This will only be populated for the user's current lobby.
+            /// </summary>
+            public IReadOnlyList<Member> Members { get; }
+
+            /// <summary>
+            /// The distance of the lobby.
+            /// </summary>
+            public float? Distance { get; }
+
+            /// <summary>
+            /// The weight of the lobby.
+            /// </summary>
+            public long? Weight { get; }
+
+            static readonly IReadOnlyDictionary<string, string> EmptyMetadata =
+                new ReadOnlyDictionary<string, string>( new Dictionary<string, string>() );
+
+            static readonly IReadOnlyList<Member> EmptyMembers = new List<Member>().AsReadOnly();
+
+            internal Lobby( SteamID steamId, ELobbyType lobbyType, int lobbyFlags, SteamID ownerSteamId, IReadOnlyDictionary<string, string> metadata,
+                int maxMembers, int numMembers, IReadOnlyList<Member> members, float? distance, long? weight )
+            {
+                SteamID = steamId;
+                LobbyType = lobbyType;
+                LobbyFlags = lobbyFlags;
+                OwnerSteamID = ownerSteamId;
+                Metadata = metadata ?? EmptyMetadata;
+                MaxMembers = maxMembers;
+                NumMembers = numMembers;
+                Members = members ?? EmptyMembers;
+                Distance = distance;
+                Weight = weight;
+            }
+
+            internal static byte[] EncodeMetadata( IReadOnlyDictionary<string, string> metadata )
+            {
+                if ( metadata == null )
+                {
+                    metadata = EmptyMetadata;
+                }
+
+                var keyValue = new KeyValue( "" );
+
+                foreach ( var entry in metadata )
+                {
+                    keyValue[ entry.Key ] = new KeyValue( null, entry.Value );
+                }
+
+                using ( var ms = new MemoryStream() )
+                {
+                    keyValue.SaveToStream( ms, true );
+
+                    var buffer = new byte[ ms.Length ];
+                    ms.Seek( 0, SeekOrigin.Begin );
+                    ms.Read( buffer, 0, buffer.Length );
+                    return buffer;
```

Comment:
This seems like a reimplementation of `MemoryStream.ToArray()`?

Reply:
Yep, good catch.

Other stuff I've been working on recently uses this sort of pattern to avoid memory allocations by using an existing buffer. However, that makes zero sense here because I've just created a brand new buffer on line 360.

I'll replace it with `buffer.ToArray()`.